### ISSUE TITLE
Omitted fields not getting treated as null when updating client's metadata.

### ISSUE THREAD

**Issue by kartikchauhan** [panva/node-oidc-provider#664]
According to section [Client Update Request](https://tools.ietf.org/html/rfc7592#section-2.2) in [RFC-7592](https://tools.ietf.org/html/rfc7592),

> Omitted fields MUST be treated as null or empty values by the server, indicating the client's request to delete them from the client's registration.

From what I understood from this statement is that when a client is updating its metadata by making a `PUT` request on the client registration URI, the client has to provide all its previous registered metadata. If the client omits any field in the request body, then that field must be treated as if the client wants to delete that from its registered metadata, right?

If I inferred it correctly, then this library should also show the same behavior. But, when I'm updating the client's metadata by omitting any field, I get the error something like _the field must be provided_.

Am I missing something here? Doesn't this violate the rules of the specification? and if so, then can the library be considered as a certified oidc-provider?

**Comment by panva** [comment]
The specification is contradicting itself. In the same section it says

```
This request MUST include all client metadata fields as returned to
   the client from a previous registration, read, or update operation.
```

So there's that.

```
Omitted fields MUST be treated as null or empty values by the server,
   indicating the client's request to delete them from the client's
   registration.  The authorization server MAY ignore any null or empty
   value in the request just as any other value.
```

> Am I missing something here? Doesn't this violate the rules of the specification?

You're not missing anything, it's an underused feature that rarely anyone uses. I'm glad you brought it up.

> can the library be considered as a certified oidc-provider?

It is not only considered, it is a certified OpenID Connect Provider. OIDC Certification has a limited scope and client registration management protocol isn't part of it.

As-is, with the current implementation, the client has no way of accidentally removing metadata just because it didn't provide it in the request and it still has means to delete what it needs by providing `null` or `""`.

As far as interoperability goes, an accepted request made to this implementation is a valid request made to the one that allows fields to be just omitted.

**Comment by kartikchauhan** [comment]
Thanks, Filip for an immediate response. 

Though, the specification says every field must be included in the statement 

> This request MUST include all client metadata fields as returned to
   the client from previous registration, read, or update operation.

But there can be a possibility the client doesn't pass all the registered fields in its request. In that case, shouldn't we treat them as null fields and delete them from the client metadata as the specification says in its statement?

> Omitted fields MUST be treated as null or empty values by the server,
   indicating the client's request to delete them from the client's
   registration.

**Comment by panva** [comment]
I'm not convinced this server implementation needs to be made ready for that case, the client can always do a GET first and then the PUT with all properties.

Nevertheless I'll relax this.

**Comment by kartikchauhan** [comment]
Thanks, Filip for asking OAuth WG.

**Comment by panva** [comment]
@kartikchauhan thanks for raising this, after clarifying with the spec authors i've removed the check and released v6.23.1
